### PR TITLE
tcp: allow exception handler to be overriden

### PR DESCRIPTION
**Pull request by pyr**
Some pipeline elements rely on exception for control flow.
This patch proposes a pluggable exception-handler which is fed
the current stream as well as the exception.

### CODE REVIEW

**Comment by vigneshsarma** [comment]
This pr might help with #529 we are facing.

File: src/aleph/tcp.clj [hunk 1]
```diff
@@ -24,14 +24,16 @@
...
     (netty/channel-inbound-handler
 
       :exception-caught
       ([_ ctx ex]
-        (when-not (instance? IOException ex)
-          (log/warn ex "error in TCP server")))
+       (if (some? exception-handler)
+         (exception-handler @in ex)
+         (when-not (instance? IOException ex)
+           (log/warn ex "error in TCP server"))))
```
Comment (vigneshsarma) [inline src/aleph/tcp.clj]: Why only log if its an IOException? 
Also, wouldnt it be better if there was a default handler? 
Something like, 
```clojure
(or exception-handler #(log/warn ex "error in TCP server"))))
```

Comment (pyr) [reply src/aleph/tcp.clj]: @vigneshsarma it was the previous behavior which I would rather leave untouched. It actually logs when the exception is not an IO one

**Comment by kachayev** [comment]
"Netty-friendly" approach would be to fire exception down the `ChannelPipeline`... In this case, `pipeline-transform` could be used to setup additional `ChannelHandler` to deal with errors. Though I understand it might be quite challenging API to adopt, I'm still curious why do we swallow unhandled exceptions without closing the connection in the first place. Seems like this might lead to a lot of problematic situations. Let me think more about this.

Having explicit `exceptions-handler` seems like a nice approach (similar to e.g. how Vert.x exposes it). Not sure if it should **only** TCP server feature... what about HTTP server? Or other servers? Maybe we need to setup `exceptions-handler` in `netty/start-server` (as a new handler in the pipeline)? 🤔

**Pull request closed without merging**